pkg/consensus/recovery_message.go

```diff
@@ -0,0 +1,172 @@
+package consensus
+
+import (
+	"github.com/CityOfZion/neo-go/pkg/io"
+	"github.com/CityOfZion/neo-go/pkg/util"
+	"github.com/pkg/errors"
+)
+
+type (
+	recoveryMessage struct {
+		PreparationHash     *util.Uint256
+		PreparationPayloads map[uint16]*preparationCompact
+		CommitPayloads      map[uint16]*commitCompact
+		ChangeViewPayloads  map[uint16]*changeViewCompact
+		PrepareRequest      *prepareRequest
+	}
+
+	changeViewCompact struct {
+		ValidatorIndex     uint16
+		OriginalViewNumber byte
+		Timestamp          uint32
+		InvocationScript   []byte
+	}
+
+	commitCompact struct {
+		ViewNumber       byte
+		ValidatorIndex   uint16
+		Signature        []byte
+		InvocationScript []byte
+	}
+
+	preparationCompact struct {
+		ValidatorIndex   uint16
+		InvocationScript []byte
+	}
+)
+
+const uint256size = 32
+
+func (m *recoveryMessage) DecodeBinary(r *io.BinReader) {
+	l := r.ReadVarUint()
+	m.ChangeViewPayloads = make(map[uint16]*changeViewCompact, l)
+	for i := uint64(0); i < l; i++ {
+		var p changeViewCompact
+		p.DecodeBinary(r)
+		m.ChangeViewPayloads[p.ValidatorIndex] = &p
+	}
+
+	var hasReq bool
+	r.ReadLE(&hasReq)
+	if hasReq {
+		m.PrepareRequest = new(prepareRequest)
+		m.PrepareRequest.DecodeBinary(r)
+	} else {
+		l := r.ReadVarUint()
+		if l != 0 {
+			if l == uint256size {
+				m.PreparationHash = new(util.Uint256)
+				r.ReadBE(m.PreparationHash[:])
+			} else {
+				r.Err = errors.New("invalid data")
+			}
+		} else {
+			m.PreparationHash = nil
+		}
+	}
+
+	l = r.ReadVarUint()
```

Comment:
lines 68-74 and 76-83 look very similar, what about generalizing and moving to some reusable function?

Reply:
We can, but in will only increase complexity IMO.

Reply:
hm
it could look like:
```
writeMessage(m.PreparationPayloads, p, reader)
writeMessage(m.CommitPayloads, p, reader)
```
which looks clear and you don't have to read logic of decoding twice

Reply:
They a both maps with different types,
I can't just write `func writeMessage(map[uint16]io.Serializable, ...`.